core/src/stage.cpp

```diff
@@ -497,6 +500,46 @@ void Generator::spawn(InterfaceState&& state, SubTrajectory&& t)
 }
 
 
+MonitoringGeneratorPrivate::MonitoringGeneratorPrivate(MonitoringGenerator *me, const std::string &name)
+   : GeneratorPrivate(me, name), monitored_(nullptr), registered_(false)
+{}
+
+MonitoringGenerator::MonitoringGenerator(const std::string &name, Stage* monitored)
+   : Generator(new MonitoringGeneratorPrivate(this, name))
+{
+	setMonitoredStage(monitored);
+}
+
+void MonitoringGenerator::setMonitoredStage(Stage* monitored)
+{
+	auto impl = pimpl();
+	if (impl->monitored_ == monitored)
+		return;
+
+	if (impl->monitored_ && impl->registered_) {
+		impl->monitored_->removeSolutionCallback(impl->cb_);
+		impl->registered_ = false;
+	}
+
+	impl->monitored_ = monitored;
+}
+
+void MonitoringGenerator::init(const moveit::core::RobotModelConstPtr& robot_model)
+{
+	Generator::init(robot_model);
+
+	auto impl = pimpl();
+	if (!impl->monitored_)
+		throw InitStageException(*this, "no monitored stage defined");
+	if (impl->monitored_->pimpl()->requiredInterface() != GENERATE)
+		throw InitStageException(*this, "monitored stage must have generator-style interface");
```

Comment:
It would be nice to avoid this restriction. I guess this is not just a matter of removing those two lines or is it?

Reply:
> This kind of stage always acts as a generator!

That's not what this checks (if I understand the code and you correctly).
The **monitoring** stage is always a generator, but I don't see what should be wrong about spawning new InterfaceStates based on solutions of any other (**monitored**) Stage.

At the moment I would like to monitor the end of solutions from a stage that modified the planning scene, so that I can spawn something based on it further back in the pipeline.

Reply:
I removed these lines locally and the pipeline behaves as expected:
for every solution generated by the (`PropagatingForward`) monitored stage the `MonitoringGenerator` can spawn new states.

Unless you see some other problem here, I would propose to remove these two lines.

Reply:
You are right. The monitored stage doesn't need to be a generator.